storage/replica.go

```diff
@@ -1660,3 +1677,55 @@ func (r *Replica) maybeAddToSplitQueue() {
 		r.store.splitQueue.MaybeAdd(r, r.store.Clock().Now())
 	}
 }
+
+// processReplicaCmd get a raft command from command queue and
+// call processRaftCommand to execute in a loop until there is
+// not any command in the queue for 5 seconds.
+func (r *Replica) processReplicaCmd() {
+	for {
+		select {
+		case cmd := <-r.cmdQueueChan:
+			err := r.processRaftCommand(cmd.idKey, cmd.index, cmd.raftCmd)
+			if cmd.callback != nil {
+				cmd.callback(err)
+			}
+
+		case <-time.After(5 * time.Second):
+			r.stopFlagLock.Lock()
+			if len(r.cmdQueueChan) == 0 {
+				r.stopFlag = true
+				r.stopFlagLock.Unlock()
+				return
+			}
+			r.stopFlagLock.Unlock()
+
+		case <-r.store.stopper.ShouldStop():
+			return
+		}
+	}
+}
+
+// addCmdQueue insert a raft command into command queue on each replica.
+// If replica's process goroutine already stopped, start a new one to get
+// and execute the command in command queue.
+func (r *Replica) addCmdQueue(idKey cmdIDKey, index uint64, raftCmd roachpb.RaftCommand, f func(error)) {
+	r.cmdQueueChan <- committedCmd{
+		idKey:    idKey,
+		index:    index,
+		raftCmd:  raftCmd,
+		callback: f,
+	}
+
+	r.stopFlagLock.Lock()
+	if r.stopFlag {
+		r.stopFlag = false
+		r.stopFlagLock.Unlock()
```

Comment:
looks like you could use an `atomic.Value` here (or even an atomically updated `int32`).